### ISSUE TITLE
Use of Groovy scripting in ElasticSearch indexing backend limits indexed property size

### ISSUE THREAD

**Issue by sharpau** [JanusGraph/janusgraph#154]
Groovy's maximum string size is 65535 characters. Because a mutation to an ES-indexed property sets that property via a Groovy script, the effective limit on ES-indexed properties is 65535 characters.

> ERROR 2017-03-07T15:08:39.485-08:00 [Grizzly(5)] com.thinkaurelius.titan.diskstorage.es.ElasticSearchIndex - Failed to execute ES query ElasticsearchIllegalArgumentException[failed to execute script]; nested: GroovyScriptCompilationException[MultipleCompilationErrorsException[startup failed:
69b79da3909d74223752de333112de783b510858: 1: String too long. The given string is 994823 Unicode code units long, but only a maximum of 65535 is allowed.
 @ line 1, column 27.
   ctx._source["Document"] = "<p>The Project Gutenberg EBook of Dracula...

This is especially unfortunate because no ES script is involved in creating a new property. So I have no problem indexing entire Project Gutenberg books in ES, as long as it's a new property and not an existing property.

I tried the workaround of deleting the property and re-adding it but this is still (apparently) a mutation to ElasticSearch.

FWIW, ElasticSearch dynamic scripting is also considered by some to be a security issue and is now off by default. It would be nice if this wasn't a requirement for JanusGraph. (As far as I can tell from the code, it is important for SET and LIST data types but is only used for SINGLE cardinality properties as a matter of convenience.)

(I observed this problem in Titan 1.0 but as far as I can tell it still applies to JanusGraph, including #79 .)

**Comment by sharpau** [comment]
Oh yeah, and the other weird part about this is that failures in index mutations don't fail the graph transaction, so ElasticSearch and the graph DB can silently get out of sync. (True in Titan, not sure if still true in JanusGraph but unless someone has specifically changed it, it probably still is.)

**Comment by MichaelMarkieta** [comment]
I can confirm the out of sync storage backend and index backend @sharpau. I just ran into that today with a very large transaction that failed during the index mutation. No rollback occurred after the index mutation failed.

```
20:00:41 ERROR org.janusgraph.graphdb.database.StandardJanusGraph  - Error while commiting index mutations for transaction [2] on index: search
org.janusgraph.core.JanusGraphException: Could not execute operation due to backend exception
       at org.janusgraph.diskstorage.util.BackendOperation.execute(BackendOperation.java:57)
       at org.janusgraph.diskstorage.indexing.IndexTransaction.flushInternal(IndexTransaction.java:134)
       at org.janusgraph.diskstorage.indexing.IndexTransaction.commit(IndexTransaction.java:115)
       at org.janusgraph.diskstorage.BackendTransaction.commitIndexes(BackendTransaction.java:140)
       at org.janusgraph.graphdb.database.StandardJanusGraph.commit(StandardJanusGraph.java:743)
       at org.janusgraph.graphdb.transaction.StandardJanusGraphTx.commit(StandardJanusGraphTx.java:1363)
       at org.janusgraph.graphdb.tinkerpop.JanusGraphBlueprintsGraph$GraphTransaction.doCommit(JanusGraphBlueprintsGraph.java:272)
       at org.apache.tinkerpop.gremlin.structure.util.AbstractTransaction.commit(AbstractTransaction.java:105)
       at org.apache.tinkerpop.gremlin.structure.Transaction$commit.call(Unknown Source)
       at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
       at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
       at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:117)
       at groovysh_evaluate.run(groovysh_evaluate:3)
       at org.codehaus.groovy.vmplugin.v7.IndyInterface.selectMethod(IndyInterface.java:232)
       at org.codehaus.groovy.tools.shell.Interpreter.evaluate(Interpreter.groovy:70)
       at org.codehaus.groovy.tools.shell.Groovysh.execute(Groovysh.groovy:190)
       at org.apache.tinkerpop.gremlin.console.GremlinGroovysh.super$3$execute(GremlinGroovysh.groovy)
       at sun.reflect.GeneratedMethodAccessor14.invoke(Unknown Source)
       at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
       at java.lang.reflect.Method.invoke(Method.java:498)
       at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
       at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
       at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1215)
       at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.invokeMethodOnSuperN(ScriptBytecodeAdapter.java:132)
       at org.apache.tinkerpop.gremlin.console.GremlinGroovysh.execute(GremlinGroovysh.groovy:72)
       at org.apache.tinkerpop.gremlin.console.Console.executeInShell(Console.groovy:370)
       at org.codehaus.groovy.vmplugin.v7.IndyInterface.selectMethod(IndyInterface.java:232)
       at org.apache.tinkerpop.gremlin.console.Console.<init>(Console.groovy:147)
       at org.codehaus.groovy.vmplugin.v7.IndyInterface.selectMethod(IndyInterface.java:232)
       at org.apache.tinkerpop.gremlin.console.Console.main(Console.groovy:401)
Caused by: org.janusgraph.diskstorage.PermanentBackendException: Permanent exception while executing backend operation IndexMutation
       at org.janusgraph.diskstorage.util.BackendOperation.executeDirect(BackendOperation.java:82)
       at org.janusgraph.diskstorage.util.BackendOperation.execute(BackendOperation.java:55)
       ... 29 more
Caused by: java.lang.OutOfMemoryError: Java heap space
       at org.elasticsearch.common.util.BigArrays.newByteArray(BigArrays.java:453)
       at org.elasticsearch.common.util.BigArrays.newByteArray(BigArrays.java:463)
       at org.elasticsearch.common.io.stream.BytesStreamOutput.<init>(BytesStreamOutput.java:60)
       at org.elasticsearch.common.io.stream.BytesStreamOutput.<init>(BytesStreamOutput.java:55)
       at org.elasticsearch.common.io.stream.BytesStreamOutput.<init>(BytesStreamOutput.java:45)
       at org.elasticsearch.common.xcontent.XContentBuilder.builder(XContentBuilder.java:77)
       at org.elasticsearch.common.xcontent.json.JsonXContent.contentBuilder(JsonXContent.java:38)
       at org.elasticsearch.common.xcontent.XContentFactory.contentBuilder(XContentFactory.java:122)
       at org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder(XContentFactory.java:49)
       at org.janusgraph.diskstorage.es.ElasticSearchIndex.getNewDocument(ElasticSearchIndex.java:492)
       at org.janusgraph.diskstorage.es.ElasticSearchIndex.mutate(ElasticSearchIndex.java:594)
       at org.janusgraph.diskstorage.indexing.IndexTransaction$1.call(IndexTransaction.java:137)
       at org.janusgraph.diskstorage.indexing.IndexTransaction$1.call(IndexTransaction.java:134)
       at org.janusgraph.diskstorage.util.BackendOperation.executeDirect(BackendOperation.java:69)
       at org.janusgraph.diskstorage.util.BackendOperation.execute(BackendOperation.java:55)
       at org.janusgraph.diskstorage.indexing.IndexTransaction.flushInternal(IndexTransaction.java:134)
       at org.janusgraph.diskstorage.indexing.IndexTransaction.commit(IndexTransaction.java:115)
       at org.janusgraph.diskstorage.BackendTransaction.commitIndexes(BackendTransaction.java:140)
       at org.janusgraph.graphdb.database.StandardJanusGraph.commit(StandardJanusGraph.java:743)
       at org.janusgraph.graphdb.transaction.StandardJanusGraphTx.commit(StandardJanusGraphTx.java:1363)
       at org.janusgraph.graphdb.tinkerpop.JanusGraphBlueprintsGraph$GraphTransaction.doCommit(JanusGraphBlueprintsGraph.java:272)
       at org.apache.tinkerpop.gremlin.structure.util.AbstractTransaction.commit(AbstractTransaction.java:105)
       at org.apache.tinkerpop.gremlin.structure.Transaction$commit.call(Unknown Source)
       at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
       at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
       at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:117)
       at groovysh_evaluate.run(groovysh_evaluate:3)
       at java.lang.invoke.LambdaForm$DMH/1811075214.invokeVirtual_L_L(LambdaForm$DMH)
       at java.lang.invoke.LambdaForm$BMH/745962066.reinvoke(LambdaForm$BMH)
       at java.lang.invoke.LambdaForm$MH/1763344271.guardWithCatch(LambdaForm$MH)
       at java.lang.invoke.LambdaForm$reinvoker/1353170030.dontInline(LambdaForm$reinvoker)
       at java.lang.invoke.LambdaForm$MH/370869802.guard(LambdaForm$MH)
```

**Comment by sjudeng** [comment]
I have a unit test that reproduces the issue with updating long text strings and also implementation that resolves the issue by splitting mutations and using an update operation for non-List/Set properties. The update is based on #79, so I was waiting on that to get merged before submitting a new PR. #79 also removes the dynamic scripting requirement on ES 5.x servers.

I'm not sure about the issue of backend rollback when index update fails. Can you create a new issue to track that?

**Comment by sjudeng** [comment]
I just added this commit to #79. That PR needs another review anyway and the update to split mutations was relatively small.

**Comment by sharpau** [comment]
Thanks a ton @sjudeng , really happy to see progress made here.

I'll leave the creation of the issue for lack of rollback to @MichaelMarkieta since I don't have a current example or test reproduction. I just noticed when looking through (Titan's) ElasticSearchIndex.java that failure is just an exception with no cleanup, beyond that (and frustration at how the DB stays broken) I don't have a lot to add.

**Comment by sjudeng** [comment]
Closing after merge of #79 which resolves this issue. Please reopen if you still have this problem.
